javatests/arcs/core/analysis/RecipeGraphFixpointIteratorTest.kt

```diff
@@ -0,0 +1,296 @@
+    is RecipeGraph.Node.Handle -> "h:${handle.name}"
+}
+
+/** A simple Fixpoint iterator for the purposes of testing. */
+class TestAnalyzer(
+    val startNames: Set<String>
+) : RecipeGraphFixpointIterator<AbstractSet<String>>(AbstractSet.getBottom<String>()) {
+    override fun getInitialValues(graph: RecipeGraph) = graph.nodes.filter {
+        startNames.contains(it.getName())
+    }.associateWith {
+        AbstractSet<String>(setOf())
+    }
+
+    override fun nodeTransfer(handle: Recipe.Handle, input: AbstractSet<String>) =
+        input.set?.let { AbstractSet<String>(it + "h:${handle.name}") } ?: input
+
+    override fun nodeTransfer(particle: Recipe.Particle, input: AbstractSet<String>) =
+        input.set?.let { AbstractSet<String>(it + "p:${particle.spec.name}") } ?: input
+
+    override fun edgeTransfer(
+        handle: Recipe.Handle,
+        particle: Recipe.Particle,
+        spec: HandleConnectionSpec,
+        input: AbstractSet<String>
+    ) = input.set?.let {
+        AbstractSet<String>(it + "h:${handle.name} -> p:${particle.spec.name}")
+    } ?: input
+
+    override fun edgeTransfer(
+        particle: Recipe.Particle,
+        handle: Recipe.Handle,
+        spec: HandleConnectionSpec,
+        input: AbstractSet<String>
+    ) = input.set?.let {
+        AbstractSet<String>(it + "p:${particle.spec.name} -> h:${handle.name}")
+    } ?: input
+}
+
+@RunWith(JUnit4::class)
+class RecipeGraphFixpointIteratorTest {
+    private val thing = Recipe.Handle("thing", Recipe.Handle.Fate.CREATE, TypeVariable("thing"))
+    private val name = Recipe.Handle("name", Recipe.Handle.Fate.CREATE, TypeVariable("name"))
+    private val readConnection = HandleConnectionSpec("r", HandleMode.Read, TypeVariable("r"))
+    private val writeConnection = HandleConnectionSpec("w", HandleMode.Write, TypeVariable("w"))
+    private val readerSpec = ParticleSpec(
+        "Reader",
+        listOf(readConnection).associateBy { it.name },
+        "ReaderLocation"
+    )
+    private val writerSpec = ParticleSpec(
+        "Writer",
+        listOf(writeConnection).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val anotherWriterSpec = ParticleSpec(
+        "AnotherWriter",
+        listOf(writeConnection).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val readerParticle = Recipe.Particle(
+        readerSpec,
+        listOf(Recipe.Particle.HandleConnection(readConnection, thing))
+    )
+    private val writerParticle = Recipe.Particle(
+        writerSpec,
+        listOf(Recipe.Particle.HandleConnection(writeConnection, thing))
+    )
+    private val anotherWriterParticle = Recipe.Particle(
+        anotherWriterSpec,
+        listOf(Recipe.Particle.HandleConnection(writeConnection, thing))
+    )
+
+    private fun createGraph(
+        name: String,
+        handles: List<Recipe.Handle>,
+        particles: List<Recipe.Particle>
+    ) = RecipeGraph(
```

Comment:
Nit: I recently learnt that the style guide says function expressions are only permitted when they fit on a single line (unless it's a curly brace block), so this should be:

```
): RecipeGraph {
    return RecipeGraph(
        Recipe(
            ...
        )
    )
}
```

Reply:
Done.